Public/Src/Cache/ContentStore/DistributedTest/Stores/GrpcCopyContentTests.cs

```diff
@@ -32,12 +32,67 @@ public class GrpcCopyContentTests : TestBase
     {
         private const int FileSize = 1000;
         private const HashType DefaultHashType = HashType.Vso0;
+        private const string LocalHost = "localhost";
         private Context _context;
+        private GrpcCopyClientCache _clientCache;
 
         public GrpcCopyContentTests()
             : base(() => new PassThroughFileSystem(TestGlobal.Logger), TestGlobal.Logger)
         {
             _context = new Context(Logger);
+            _clientCache = new GrpcCopyClientCache(_context);
+        }
+
+        [Fact]
+        public void DuplicateClientsAreTheSameObject()
+        {
+            using (var client1 = _clientCache.Create(LocalHost, 10, true))
+            using (var client2 = _clientCache.Create(LocalHost, 10, true))
+            {
+                Assert.Same(client1, client2);
+            }
+        }
+
+        [Fact]
+        public void ValidateBackgroundCleanup()
+        {
+            var key = new GrpcCopyClientKey(LocalHost, 11, true);
+            using (var client = _clientCache.Create(key.Host, key.GrpcPort, key.UseCompression))
+            { 
+                client._lastUseTime = DateTime.UtcNow - TimeSpan.FromHours(2);
+            }
+
+            _clientCache.StartBackgroundCleanup();
+            var endTime = DateTime.UtcNow + TimeSpan.FromMinutes(1);
+            while (DateTime.UtcNow < endTime)
+            {
+                if (!_clientCache._clientDict.TryGetValue(key, out GrpcCopyClient foundClient))
+                {
+                    return;
+                }
+
+                Task.Delay(1000).GetAwaiter().GetResult();

```

Comment:
We should switch to IClock and not to wait for a minute for this.

The current approach is more brittle and will take way more time than needed to run the tests.

Reply:
Because I think that IClock is messy. Would switching to a significantly shorter delay like 100ms be better?

---
In reply to: [290518437](https://github.com/microsoft/BuildXL/pull/391#discussion_r290518437) [](ancestors = 290518437)